ipq807x: prpl-haze: enable LED driver on device tree

CheWei-Chien opened the pull request:
Enable prpl-haze LED driver LP5562 on device tree.

You also need to set:
CONFIG_LEDS_LP5562=y
CONFIG_LEDS_LP55XX_COMMON=y
in kernel config

Signed-off-by: CheWei Chien <chewei.chien@wnc.com.tw>

robimarko commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8072-haze.dts:
```diff
@@ -78,6 +78,13 @@
 		};
 	};
+
+	i2c_0_pins: i2c-0-state {
```
This is not I2C0, so rename it accordingly.

robimarko commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8072-haze.dts:
```diff
@@ -303,3 +310,39 @@
+	status = "ok";
+
+	led-controller@30 {
+		compatible = "ti,lp5562";
```
How is the driver kmod included on this device?
I dont see it being added to the default packages

robimarko commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8072-haze.dts:
```diff
@@ -303,3 +310,39 @@
+			max-cur = /bits/ 8 <0x60>;
+			color = <LED_COLOR_ID_RED>;
+			reg = <0>;
+		};
```
Add newline after each subnode

CheWei-Chien commented:
Thank for your review!
I do some fix in new commit.
I think the driver kmod should enable on openwrt/target/linux/generic/config-6.1 if possible.

ynezz commented:
> I think the driver kmod should enable on openwrt/target/linux/generic/config-6.1 if possible.

Although we've already following prior example in the [ipq40xx target](https://github.com/openwrt/openwrt/blob/main/target/linux/ipq40xx/config-6.1#L244-L246):

```
target/linux/ipq40xx/config-6.1:CONFIG_LEDS_LP5523=y
target/linux/ipq40xx/config-6.1:CONFIG_LEDS_LP5562=y
target/linux/ipq40xx/config-6.1:CONFIG_LEDS_LP55XX_COMMON=y
```

I think, that the correct approach here would be to add new kernel module package `KernelPackage/leds-lp5562` into `package/kernel/linux/modules/leds.mk` and then add it to `DEVICE_PACKAGES` kernel module list in the image definition.

According to the [DTS checklist](https://openwrt.org/submitting-patches#dts_checklist)
> 4. If it's possible try to dedicate some of the LEDs for system status indication in [diag.sh](https://github.com/openwrt/openwrt/blob/master/package/base-files/files/etc/diag.sh) ([example](https://github.com/openwrt/openwrt/blob/master/target/linux/ath79/dts/qca9533_tplink_tl-wr841-v11.dtsi#L7))

It makes me wonder if it would be possible to define such LED aliases as well?

robimarko commented:
Please squash the commits, no need for those.

And no, it should not be enabled in the generic config, please add the required kmod as default package for this device

ynezz commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8072-haze.dts:
```diff
@@ -303,3 +310,41 @@
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		status_red: chan@0 {
```
Nitpick, there seems to be a convention to name the aliases in the following way

```suggestion
		led_system_red: chan@0 {
```

ynezz commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8072-haze.dts:
```diff
@@ -303,3 +310,41 @@
+			reg = <0>;
+		};
+
+		status_green: chan@1 {
```
Nitpick, there seems to be a convention to name the aliases in the following way

```suggestion
		led_system_green: chan@1 {
```

ynezz commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8072-haze.dts:
```diff
@@ -303,3 +310,41 @@
+			reg = <1>;
+		};
+
+		status_blue: chan@2 {
```
Nitpick, there seems to be a convention to name the aliases in the following way

```suggestion
		led_system_blue: chan@2 {
```

ynezz commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8072-haze.dts:
```diff
@@ -303,3 +310,41 @@
+			max-cur = /bits/ 8 <0x60>;
+			color = <LED_COLOR_ID_BLUE>;
+			reg = <2>;
+		};
```
I don't have access to the device now, so can't check it myself, but I'm wondering if the 3rd/white LED channel was left intentionally or its a mistake?

CheWei-Chien replied:
The 3rd/white LED channel is broken at the physical layer, so I didn't set it.

CheWei-Chien commented:
Thank for yours reviews!
I squash the commits, add new kernel module package for LED driver and included it on default package for HAZE.
If have any mistake please let me know, thank you.

ynezz commented on target/linux/qualcommax/image/ipq807x.mk:
```diff
@@ -195,7 +195,7 @@ define Device/prpl_haze
 	SOC := ipq8072
 	DEVICE_PACKAGES += ath11k-firmware-qcn9074 ipq-wifi-prpl_haze kmod-ath11k-pci \
-		mkf2fs f2fsck kmod-fs-f2fs
+		mkf2fs f2fsck kmod-fs-f2fs kmod-leds-lp55xx-common kmod-leds-lp5562
```
```suggestion
		mkf2fs f2fsck kmod-fs-f2fs kmod-leds-lp5562
```

`kmod-leds-lp5562` already depends on `kmod-leds-lp55xx-common`

ynezz commented on package/kernel/linux/modules/leds.mk:
```diff
@@ -261,3 +261,37 @@ define KernelPackage/input-leds/description
 $(eval $(call KernelPackage,input-leds))
+
+
+define KernelPackage/leds-lp55xx-common
```
This kmods needs to be a separate commit, see for example 77a14795dfd5488d62ec5f5cf04b7bd7a8c278ee or generally `git log package/kernel/linux/modules/leds.mk` for prior examples.

ynezz commented:
> Thank for yours reviews! I squash the commits, add new kernel module package for LED driver and included it on default package for HAZE. If have any mistake please let me know, thank you.

You've probably missed [following part](https://github.com/openwrt/openwrt/pull/14290#issuecomment-1873743943):

> According to the [DTS checklist](https://openwrt.org/submitting-patches#dts_checklist)
> 
> If it's possible try to dedicate some of the LEDs for system status indication in [diag.sh](https://github.com/openwrt/openwrt/blob/master/package/base-files/files/etc/diag.sh) ([example](https://github.com/openwrt/openwrt/blob/master/target/linux/ath79/dts/qca9533_tplink_tl-wr841-v11.dtsi#L7))
> 
> It makes me wonder if it would be possible to define such LED aliases as well?

So either add the LED aliases (preferred) so we get the nice visual LED status feedback during various system phases (booting, upgrading etc.) or you should actually remove the `led_system_*:` labels as there are no users for them.

From https://openwrt.org/docs/guide-developer/device-support-policies#leds

>  The DT label (not the label property) should be created from the node name by adding a led_ prefix, because the DT label will be used in global context. Only add a DT label if you use it for something, otherwise just omit it.

CheWei-Chien committed:
kernel: add kmod-leds-lp5562 and kmod-leds-lp55xx-common

Add kernel module for lp5562 LED driver.
The kmod-leds-lp5562 depends on kmod-leds-lp55xx-common.

Signed-off-by: CheWei Chien <chewei.chien@wnc.com.tw>

CheWei-Chien committed:
ipq807x: prpl-haze: enable LED driver on device tree

Enable LED driver LP5562 on HAZE device tree and include its kernel
module package on default package for HAZE.

Signed-off-by: CheWei Chien <chewei.chien@wnc.com.tw>

ynezz reviewed (approved):
Thanks! Tested, works, LGTM
